lib/sshutils/sftp/sftp_test.go

```diff
@@ -0,0 +1,435 @@
+				if strings.HasSuffix(file, string(filepath.Separator)) {
+					createDir(t, tempDir, file)
+				} else {
+					createFile(t, tempDir, file)
+				}
+			}
+			for i := range tt.srcPaths {
+				tt.srcPaths[i] = filepath.Join(tempDir, tt.srcPaths[i])
+			}
+			tt.dstPath = filepath.Join(tempDir, tt.dstPath)
+
+			ctx := context.Background()
+			cfg := CreateUploadConfig(tt.srcPaths, tt.dstPath, tt.opts)
+			// use all local filesystems to avoid SSH overhead
+			cfg.dstFS = &localFS{}
+			cfg.initFS(ctx, nil)
+
+			err := cfg.transfer(ctx)
+			if tt.expectedErr == "" {
+				require.NoError(t, err)
+				checkTransfer(t, tt.opts.PreserveAttrs, tt.dstPath, tt.srcPaths...)
+			} else {
+				require.EqualError(t, err, fmt.Sprintf(tt.expectedErr, tempDir))
+			}
+		})
+	}
+}
+
+func TestDownload(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		name        string
+		srcPath     string
+		dstPath     string
+		opts        Options
+		files       []string
+		expectedErr string
+	}{
+		{
+			name:    "one file",
+			srcPath: "file",
+			dstPath: "copied-file",
+			opts: Options{
+				PreserveAttrs: true,
+			},
+			files: []string{
+				"file",
+			},
+		},
+		{
+			name:    "one dir",
+			srcPath: "src/",
+			dstPath: "dst/",
+			opts: Options{
+				PreserveAttrs: true,
+				Recursive:     true,
+			},
+			files: []string{
+				"src/",
+				"s/file",
+				"s/r/",
+				"s/r/file",
+				"s/r/c/",
+				"s/r/c/file",
+				"dst/",
+			},
+		},
+		{
+			name:    "src dir with recursive not passed",
+			srcPath: "src/",
+			dstPath: "dst/",
+			files: []string{
+				"src/",
+			},
+			expectedErr: `"%s/src" is a directory, but the recursive option was not passed`,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// create necessary files
+			tempDir := t.TempDir()
+			for _, file := range tt.files {
+				// if path ends in slash, create dir
+				if strings.HasSuffix(file, string(filepath.Separator)) {
+					createDir(t, tempDir, file)
+				} else {
+					createFile(t, tempDir, file)
+				}
+			}
+			tt.srcPath = filepath.Join(tempDir, tt.srcPath)
+			tt.dstPath = filepath.Join(tempDir, tt.dstPath)
+
+			ctx := context.Background()
+			cfg := CreateDownloadConfig(tt.srcPath, tt.dstPath, tt.opts)
+			// use all local filesystems to avoid SSH overhead
+			cfg.srcFS = &localFS{}
+			cfg.initFS(ctx, nil)
+
+			err := cfg.transfer(ctx)
+			if tt.expectedErr == "" {
+				require.NoError(t, err)
+				checkTransfer(t, tt.opts.PreserveAttrs, tt.dstPath, tt.srcPath)
+			} else {
+				require.EqualError(t, err, fmt.Sprintf(tt.expectedErr, tempDir))
+			}
+		})
+	}
+}
+
+func createFile(t *testing.T, rootDir, path string) {
+	dir := filepath.Dir(path)
+	if dir != path {
+		createDir(t, rootDir, dir)
+	}
+
+	// ensure we can read and write to this file, but give random permissions
+	// to everyone to test that file permissions are retained when transferring
+	mode := os.FileMode(0o660 + rand.Intn(8))
```

Comment:
I don't think that using random permissions is a good idea. Why just not change permissions on the source and compare it with the destination? This also makes the test non-deterministic.

Reply:
Changed to use static permissions, this should test it well enough.